## Bibliotecas

In [5]:
%pip install pandas
%pip install plotly_express
%pip install missingno
%pip install scipy
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip



[notice] A new release of pip is available: 23.1.2 -> 23.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


## Importações

In [6]:
from math import sqrt
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

c:\Users\Inteli\AppData\Local\Programs\Python\Python311\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


## Carregamento de Dados

In [7]:
# Importação dos dados da mobly sem codificação
df_no_codification = pd.read_csv("../documentos/outros/base_de_dados/mobly_data_no_codification.csv", sep=",")
df_no_codification

date weekday_name                     sku  unit_price  \
0       2020-01-01    Wednesday  MO173UP38VJHMOB-414464     1099.99   
1       2020-01-01    Wednesday  OL487CL08DELMOB-384528      413.99   
2       2020-01-01    Wednesday  RU994TA18APHMOB-575967      949.99   
3       2020-01-01    Wednesday  ES745UP35LGIMOB-669318     2949.99   
4       2020-01-01    Wednesday  OR067MA29HJAMOB-492877      657.99   
...            ...          ...                     ...         ...   
780105  2023-07-03       Monday  CI381BE06YKZMOB-178176      504.88   
780106  2023-07-03       Monday  HR667UP36BSLMOB-988738     1717.95   
780107  2023-07-03       Monday  DE626CL54BLLMOB-798176     1659.88   
780108  2023-07-03       Monday  IM630CH36ZRVMOB-864724      909.96   
780109  2023-07-03       Monday  MO173CH87TWQMOB-519074      619.98   

       shipment_type           anchor_category product_department  \
0       crossdocking                     Sofas               Keva   
1       crossdocking  Dressers and Nightstands             Móveis   
2       crossdocking               Dining Sets             Móveis   
3       crossdocking                     Sofas             Móveis   
4       crossdocking                Mattresses             Móveis   
...              ...                       ...                ...   
780105  crossdocking         Beds & Headboards             Móveis   
780106  crossdocking                  Sofabeds             Móveis   
780107  crossdocking                   Closets             Móveis   
780108  crossdocking             Dining Chairs             Móveis   
780109       próprio                Bar Stools               Keva   

       product_category origin_country process_costing sku_color  \
0         Sala de Estar       Nacional             yes     cinza   
1                Quarto       Nacional              no  carvalho   
2        Sala de Jantar       Nacional              no    imbuia   
3         Sala de Estar       Nacional              no    marfim   
4                Quarto       Nacional              no     preto   
...                 ...            ...             ...       ...   
780105           Quarto       Nacional              no    branco   
780106    Sala de Estar       Nacional              no     cinza   
780107           Quarto       Nacional              no    marrom   
780108   Sala de Jantar       Nacional              no      bege   
780109   Sala de Jantar      Importado              no     preto   

            price_status  winning_price  revenue  items_sold  \
0           Sem Registro            NaN     0.00           0   
1           Sem Registro            NaN     0.00           0   
2           Sem Registro            NaN   981.02           1   
3           Sem Registro            NaN     0.00           0   
4               Perdendo         529.00     0.00           0   
...                  ...            ...      ...         ...   
780105          Perdendo         339.00     0.00           0   
780106  Único Disponível        1717.95     0.00           0   
780107          Perdendo        1093.68     0.00           0   
780108          Perdendo         899.96     0.00           0   
780109      Sem Registro            NaN     0.00           0   

        avg_website_visits_last_week  stock_qty  month  day  year  
0                          17.285714          0      1    1  2020  
1                          48.857143         56      1    1  2020  
2                          22.714286          0      1    1  2020  
3                           2.800000          0      1    1  2020  
4                           0.000000          0      1    1  2020  
...                              ...        ...    ...  ...   ...  
780105                     41.000000         61      7    3  2023  
780106                    154.142857          1      7    3  2023  
780107                    138.000000         16      7    3  2023  
780108                     32.714286         20      7    3  2023 

## Preenchimento da coluna `winning_price`

A base de dados fornecida pela Mobly possui uma grande quantidade de dados nulos na coluna `winning_price`, como é possível observar no código abaixo:

In [8]:
# Verifica a presença de dados nulos na coluna "winning_price"
df_no_codification["winning_price"].isnull().sum()

621354

Dessa forma, na 3° sprint, o parceiro disponibilizou uma nova base de dados, a qual estamos chamando de `winning_price.csv`, que contém apenas os dados relacionados à coluna `winning_price`. O parceiro disponibilizou tal base de dados devido ao pedido da turma como um todo, haja vista que tal coluna possui muitos dados faltantes. 
Ademais, é válido ressaltar que o parceiro foi quem apontou a importância da permanência de tal coluna na base de dados que será utilizada para treinar o modelo preditivo que está sendo desenvolvido. 

Dessa forma, realizaremos o preenchimento dessa coluna abaixo:

In [9]:
# Importa os dados relacionados à coluna "winning_price"
df_winning_price = pd.read_csv("../documentos/outros/base_de_dados/winning_price.csv", encoding='latin-1')
df_winning_price

date            nk_sku_simple      price_status  winning_price  \
0       2023-05-09   HR667UP36BSLMOB-988738  Único Disponível        1717.95   
1       2023-05-09   CI381AC17VYGMOB-967333          Perdendo         152.91   
2       2023-05-08   NE528TA72UWDMOB-984126          Perdendo         346.62   
3       2023-05-07  GU364AC33MYSMOB-1065519          Perdendo        3365.10   
4       2023-05-07  GU364MA54ECFMOB-1007582  Único Disponível         899.95   
...            ...                      ...               ...            ...   
163434  2021-09-26    MA566UP11IOMMOB-77265          Perdendo         780.35   
163435  2021-09-12    MA566UP98YYBMOB-52950          Perdendo         678.96   
163436  2021-09-26   LI582UP10OKJMOB-116816          Perdendo         693.83   
163437  2021-10-03    PH779UP72MTRMOB-80104         Empatando         306.99   
163438  2022-02-20   LI582UP16OKDMOB-116810          Perdendo         846.12   

        americanas_price  casas_bahia_price  madeira_madeira_price  \
0                    NaN                NaN                    NaN   
1                 169.89              207.9                  229.0   
2                 346.62                NaN                  417.2   
3                    NaN                NaN                    NaN   
4                    NaN                NaN                    NaN   
...                  ...                ...                    ...   
163434            780.35                NaN                    NaN   
163435            678.96                NaN                    NaN   
163436            693.83                NaN                    NaN   
163437               NaN                NaN                    NaN   
163438               NaN                NaN                    NaN   

        extra_price  tricae_price  magazine_luiza_price  mobly_price  
0               NaN           NaN                   NaN      1717.95  
1             207.9           NaN                168.99       199.96  
2               NaN           NaN                346.68       439.96  
3               NaN           NaN                   NaN      3499.99  
4               NaN           NaN                   NaN       899.95  
...             ...           ...                   ...          ...  
163434          NaN           NaN                   NaN       928.99  
163435          NaN           NaN                   NaN       778.99  
163436          NaN           NaN                   NaN       825.99  
163437          NaN           NaN                   NaN       306.99  
163438          NaN           NaN                   NaN       852.99  

[163439 rows x 11 columns]

In [10]:
# Renomeia a coluna "nk_sku_simple" para "sku" no dataset "df_winning_price"
df_winning_price.rename(columns={'nk_sku_simple': 'sku'}, inplace=True)
df_winning_price

date                      sku      price_status  winning_price  \
0       2023-05-09   HR667UP36BSLMOB-988738  Único Disponível        1717.95   
1       2023-05-09   CI381AC17VYGMOB-967333          Perdendo         152.91   
2       2023-05-08   NE528TA72UWDMOB-984126          Perdendo         346.62   
3       2023-05-07  GU364AC33MYSMOB-1065519          Perdendo        3365.10   
4       2023-05-07  GU364MA54ECFMOB-1007582  Único Disponível         899.95   
...            ...                      ...               ...            ...   
163434  2021-09-26    MA566UP11IOMMOB-77265          Perdendo         780.35   
163435  2021-09-12    MA566UP98YYBMOB-52950          Perdendo         678.96   
163436  2021-09-26   LI582UP10OKJMOB-116816          Perdendo         693.83   
163437  2021-10-03    PH779UP72MTRMOB-80104         Empatando         306.99   
163438  2022-02-20   LI582UP16OKDMOB-116810          Perdendo         846.12   

        americanas_price  casas_bahia_price  madeira_madeira_price  \
0                    NaN                NaN                    NaN   
1                 169.89              207.9                  229.0   
2                 346.62                NaN                  417.2   
3                    NaN                NaN                    NaN   
4                    NaN                NaN                    NaN   
...                  ...                ...                    ...   
163434            780.35                NaN                    NaN   
163435            678.96                NaN                    NaN   
163436            693.83                NaN                    NaN   
163437               NaN                NaN                    NaN   
163438               NaN                NaN                    NaN   

        extra_price  tricae_price  magazine_luiza_price  mobly_price  
0               NaN           NaN                   NaN      1717.95  
1             207.9           NaN                168.99       199.96  
2               NaN           NaN                346.68       439.96  
3               NaN           NaN                   NaN      3499.99  
4               NaN           NaN                   NaN       899.95  
...             ...           ...                   ...          ...  
163434          NaN           NaN                   NaN       928.99  
163435          NaN           NaN                   NaN       778.99  
163436          NaN           NaN                   NaN       825.99  
163437          NaN           NaN                   NaN       306.99  
163438          NaN           NaN                   NaN       852.99  

[163439 rows x 11 columns]

In [11]:
# Cria um dicionário mapeando SKU para winning_price em "df_winning_price"
sku_to_winning_price = df_winning_price.set_index('sku')['winning_price'].to_dict()

# Preenche os valores nulos em df_no_date com base no dicionário
df_no_codification['winning_price'].fillna(df_no_codification['sku'].map(sku_to_winning_price), inplace=True)

In [12]:
# Verifica se há dados nulos na coluna 'winning_price'
df_no_codification["winning_price"].isnull().sum()

41170

In [13]:
# Verifica quais são os SKUs que ainda possuem valores nulos na coluna 'winning_price'
nulls_skus = df_no_codification[df_no_codification["winning_price"].isnull()]["sku"].unique()
nulls_skus

array(['AC967UP19DSYMOB-181803', 'FR314UP49QNAMOB-360482',
       'FR314UP60QMPMOB-360471', 'MO173UP29MEGMOB-390606',
       'CR897TA86FLTMOB-474214', 'SA756CL33CRQMOB-108841',
       'FR314UP50QMZMOB-360481', 'SO964UP01PVIMOB-498931',
       'SO964UP15PDGMOB-602899', 'SO964UP05PVEMOB-498927',
       'FR314UP42QNHMOB-360489', 'AC967UP21DSWMOB-181801',
       'SO964UP12PDJMOB-602902', 'FR314UP56QMTMOB-360475',
       'FR314UP38QNLMOB-360493', 'MO173UP07TBWMOB-430516',
       'ZA259CH85CRQMOB-489753', 'AC967UP22DSVMOB-181800',
       'MO173RA60LOHMOB-548353', 'AC967UP20DSXMOB-181802',
       'FR314UP55XPOMOB-317960', 'MO173AC71YHKMOB-643614',
       'CI381TA40NAPMOB-531911', 'FR314UP52QMXMOB-360479',
       'LI729RA15TRKMOB-657768', 'MA019CL79POYMOB-312360',
       'ZA259TA07OBKMOB-311332', 'FR314UP58QMRMOB-360473',
       'RU994CL96WDVMOB-555451', 'ZA259CH19OAYMOB-311320',
       'ZA259CH86CRPMOB-489752', 'FR314UP79QVSMOB-690215',
       'CR214TA31RNSMOB-774543', 'MO173CL99SBWMOB-101696

A partir da identificação dos SKU's que ainda possuem dados nulos, levamos para o parceiro uma dúvida relacionada à ação que deveríamos tomar sobre esses produtos, haja vista que eles **não estavam na base de `winning_price` que foi enviada pelo parceiro** e só estava na **nossa base**. Sendo assim, enviamos tal dúvida para o parceiro Mobly por intermédio da orientadora e como resposta, obtivemos que a **maioria desses SKU's são exclusivos da Mobly**, ou seja, não possuem um preço nas concorrências e, consequentemente, não há dados para serem preenchidos na coluna `winning_price`. 
&nbsp;

Com isso, decidimos preencher os valores nulos de `winning_price` dos 35 SKU's listados acima por meio da cópia do `unit_price` naquela mesma data de venda, haja vista que a maioria são exclusivos e não irão possuir o valor original da coluna `winning_price`. Aqueles que não são exclusivos, como são dados que nem o parceiro e nós temos, também substuímos pelo `unit_price`, de modo a tornar os dados um pouco mais equilibrados e sem a presença de dados nulos. Sendo assim, esse preenchimento será realizado logo a seguir:

In [14]:
# Realiza uma cópia do DataFrame "df_no_codification"
df_updated = df_no_codification.copy()

# Atualiza os valores nulos em "winning_price" na cópia com "unit_price" correspondente
df_updated.query('winning_price.isnull()', inplace=True)
df_updated['winning_price'] = df_updated['unit_price']

# Agrupa os registros tratados com o DataFrame original
df_no_codification = df_no_codification.combine_first(df_updated)

In [15]:
# Verifica se ainda há dados nulos
df_no_codification["winning_price"].isnull().sum()

0

Através do código acima é possível perceber que não há mais a presença de dados nulos na coluna `winning_price`. 
&nbsp;

Sendo assim, uma forma de verificarmos se realmente o preenchimento dos valores nulos de `winning_price` pelo valor de `unit_price` ocorreu, executamos o seguinte código:

In [16]:
# Verifica o preenchimento dos valores nulos da "winning_price" pelo valor da "unit_price"
value_sought = 'MO173UP29MEGMOB-390606'
result = df_no_codification[df_no_codification['sku'] == value_sought]
result

date weekday_name                     sku  unit_price  \
31      2020-01-01    Wednesday  MO173UP29MEGMOB-390606     2299.99   
482     2020-01-02     Thursday  MO173UP29MEGMOB-390606     2299.99   
860     2020-01-03       Friday  MO173UP29MEGMOB-390606     2299.99   
1238    2020-01-04     Saturday  MO173UP29MEGMOB-390606     2199.99   
1618    2020-01-05       Sunday  MO173UP29MEGMOB-390606     2199.99   
...            ...          ...                     ...         ...   
776474  2023-06-29     Thursday  MO173UP29MEGMOB-390606     4014.95   
777278  2023-06-30       Friday  MO173UP29MEGMOB-390606     4014.95   
778021  2023-07-01     Saturday  MO173UP29MEGMOB-390606     4014.95   
778803  2023-07-02       Sunday  MO173UP29MEGMOB-390606     4014.95   
779567  2023-07-03       Monday  MO173UP29MEGMOB-390606     4014.95   

       shipment_type anchor_category product_department product_category  \
31      crossdocking           Sofas               Keva    Sala de Estar   
482     crossdocking           Sofas               Keva    Sala de Estar   
860     crossdocking           Sofas               Keva    Sala de Estar   
1238    crossdocking           Sofas               Keva    Sala de Estar   
1618    crossdocking           Sofas               Keva    Sala de Estar   
...              ...             ...                ...              ...   
776474  crossdocking           Sofas               Keva    Sala de Estar   
777278  crossdocking           Sofas               Keva    Sala de Estar   
778021  crossdocking           Sofas               Keva    Sala de Estar   
778803  crossdocking           Sofas               Keva    Sala de Estar   
779567  crossdocking           Sofas               Keva    Sala de Estar   

       origin_country process_costing sku_color  price_status  winning_price  \
31           Nacional              no     cinza  Sem Registro        2299.99   
482          Nacional              no     cinza  Sem Registro        2299.99   
860          Nacional              no     cinza  Sem Registro        2299.99   
1238         Nacional              no     cinza  Sem Registro        2199.99   
1618         Nacional              no     cinza  Sem Registro        2199.99   
...               ...             ...       ...           ...            ...   
776474       Nacional              no     cinza  Sem Registro        4014.95   
777278       Nacional              no     cinza  Sem Registro        4014.95   
778021       Nacional              no     cinza  Sem Registro        4014.95   
778803       Nacional              no     cinza  Sem Registro        4014.95   
779567       Nacional              no     cinza  Sem Registro        4014.95   

        revenue  items_sold  avg_website_visits_last_week  stock_qty  month  \
31         0.00           0                     41.428571          0      1   
482        0.00           0                     41.571429          0      1   
860     4490.98           2                     40.714286          0      1   
1238       0.00           0                     41.142857          0      1   
1618    2257.53           1                     45.000000          0      1   
...         ...         ...                           ...        ...    ...   
776474     0.00           0                     27.714286          1      6   
777278  3849.30           1                     28.000000          0      6   
778021  3457.60           1                     30.714286          0      7   
778803     0.00           0                     31.571429          0      7   
779567     0.00           0                     30.000000          0      7   

        day  year  
31        1  2020  
482       2  2020  
860       3  2020  
1238      4  2020  
1618      5  2020  
...     ...   ...  
776474   29  2023  
777278   30  2023  
778021    1  2023  
778803    2  2023  
779567    3  2023  

[1277 rows x 20 columns]

Verificando o dataframe acima é possível observar que, na linha 31, o `unit_price` do SKU MO173UP29MEGMOB-390606 é **2299.99** e o seu `winning_price` também está valendo **2299.99**. É possível observar isso na linha 776474, onde o `unit_price` do SKU já citado é **4014.95** e o `winning_price` dessa coluna também vale esse valor. Sendo assim, a substituição dos valores da coluna `winning_price` que estavam nulos para o valor de `unit_price` correspondente ao SKU ocorreu com sucesso.

### Codificação das variáveis categóricas após o preenchimento da coluna `winning_price`

Como o preenchimento da coluna `winning_price` necessitou da base de dados fornecida pela Mobly sem a codificação das variáveis categóricas, é necessário codificá-las novamente, de modo que os dados possam ser utilizados pelo modelo preditivo que está sendo desenvolvido:

Variáveis categóricas: `weekday_name`, `sku`, `shipment_type`, `origin_country`, `process_costing`, `sku_color`, `price_status`, `product_category`.

In [17]:
# Importa o LabelEncoder
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()

In [18]:
# Faz uma cópia do dataframe sem codificação 
df_codification = df_no_codification.copy()

# Codifica as colunas categóricas que possuem múltiplos valores com LabelEncoder
columns_to_encode = ['weekday_name', 'sku_color', 'anchor_category', 'product_category', 'sku', 'product_department', 'price_status']
df_codification[columns_to_encode] = df_codification[columns_to_encode].apply(le.fit_transform)

# Codifica as colunas categóricas que possuem valores binários com Map
df_codification['shipment_type'] = df_codification['shipment_type'].map({'crossdocking': 0, 'próprio': 1})
df_codification['origin_country'] = df_codification['origin_country'].map({'Nacional': 0, 'Importado': 1})
df_codification['process_costing'] = df_codification['process_costing'].map({'yes': 0, 'no': 1})


In [19]:
# Verifica se a codificação foi realizada corretamente
df_codification

date  weekday_name  sku  unit_price  shipment_type  \
0       2020-01-01             6  549     1099.99              0   
1       2020-01-01             6  669      413.99              0   
2       2020-01-01             6  745      949.99              0   
3       2020-01-01             6  129     2949.99              0   
4       2020-01-01             6  697      657.99              0   
...            ...           ...  ...         ...            ...   
780105  2023-07-03             1   48      504.88              0   
780106  2023-07-03             1  182     1717.95              0   
780107  2023-07-03             1  109     1659.88              0   
780108  2023-07-03             1  184      909.96              0   
780109  2023-07-03             1  446      619.98              1   

        anchor_category  product_department  product_category  origin_country  \
0                    24                   3                11               0   
1                    13                   4                 9               0   
2                    10                   4                12               0   
3                    24                   4                11               0   
4                    17                   4                 9               0   
...                 ...                 ...               ...             ...   
780105                2                   4                 9               0   
780106               23                   4                11               0   
780107                5                   4                 9               0   
780108                9                   4                12               0   
780109                1                   3                12               1   

        process_costing  sku_color  price_status  winning_price  revenue  \
0                     0         12             3        1488.07     0.00   
1                     1          9             3         392.58     0.00   
2                     1         17             3        1110.69   981.02   
3                     1         21             3        3457.04     0.00   
4                     1         31             2         529.00     0.00   
...                 ...        ...           ...            ...      ...   
780105                1          6             2         339.00     0.00   
780106                1         12             4        1717.95     0.00   
780107                1         22             2        1093.68     0.00   
780108                1          5             2         899.96     0.00   
780109                1         31             3         659.98     0.00   

        items_sold  avg_website_visits_last_week  stock_qty  month  day  year  
0                0                     17.285714          0      1    1  2020  
1                0                     48.857143         56      1    1  2020  
2                1                     22.714286          0      1    1  2020  
3                0                      2.800000          0      1    1  2020  
4                0                      0.000000          0      1    1  2020  
...            ...                           ...        ...    ...  ...   ...  
780105           0                     41.000000         61      7    3  2023  
780106           0                    154.142857          1      7    3  2023  
780107           0                    138.000000         16      7    3  2023  
780108           0                     32.714286         20      7    3  2023  
780109           0                     19.000000        136      7    3  2023  

[780110 rows x 20 columns]

## Adição das variáveis externas de `dólar` e `empregabilidade`

Como o dataframe utilizado para preencher a variável `winning_price` é um dataframe inicial, ele não continha as variáveis externas que já tínhamos colocado no dataframe no caderno _Modelagem_. Sendo assim, iremos atribuir novamente tais variáveis no dataframe atual que possui a coluna `winning_price` prenchida:

In [20]:
# Importa o CSV que contém os dados da variável externa de dólar
df_dollar = pd.read_csv("../documentos/outros/base_de_dados/dollar.csv")
df_dollar

year_dollar  month_dollar  value_dollar
0          2020             1        4.0949
1          2020             2        4.1622
2          2020             3        4.3163
3          2020             4        4.7362
4          2020             5        5.2579
5          2020             6        5.8229
6          2020             7        5.1883
7          2020             8        5.3491
8          2020             9        5.3852
9          2020            10        5.2728
10         2020            11        5.6172
11         2020            12        5.4854
12         2021             1        5.0968
13         2021             2        5.2714
14         2021             3        5.3815
15         2021             4        5.6296
16         2021             5        5.6234
17         2021             6        5.2701
18         2021             7        5.0874
19         2021             8        5.1000
20         2021             9        5.2474
21         2021            10        5.2576
22         2021            11        5.4510
23         2021            12        5.4199
24         2022             1        5.7127
25         2022             2        5.5349
26         2022             3        5.1881
27         2022             4        5.1314
28         2022             5        4.7158
29         2022             6        5.1075
30         2022             7        5.1119
31         2022             8        5.4014
32         2022             9        5.0925
33         2022            10        5.2211
34         2022            11        5.2824
35         2022            12        5.3013
36         2023             1        5.3183
37         2023             2        5.1146
38         2023             3        5.2243
39         2023             4        5.2981
40         2023             5        4.9455
41         2023             6        4.9103
42         2023             7        4.8222
43         2023             8        0.0000
44         2023             9        0.0000
45         2023            10        0.0000
46         2023            11        0.0000
47         2023            12        0.0000

In [21]:
# Importa o CSV que contém os dados da variável externa de empregabilidade
df_employability = pd.read_csv("../documentos/outros/base_de_dados/employability.csv")
df_employability

year_employability  month_employability  rate_employability
0                 2020                    1                12.4
1                 2020                    2                12.4
2                 2020                    3                12.4
3                 2020                    4                13.6
4                 2020                    5                13.6
5                 2020                    6                13.6
6                 2020                    7                14.9
7                 2020                    8                14.9
8                 2020                    9                14.9
9                 2020                   10                14.2
10                2020                   11                14.2
11                2020                   12                14.2
12                2021                    1                14.9
13                2021                    2                14.9
14                2021                    3                14.9
15                2021                    4                14.2
16                2021                    5                14.2
17                2021                    6                14.2
18                2021                    7                12.6
19                2021                    8                12.6
20                2021                    9                12.6
21                2021                   10                11.1
22                2021                   11                11.1
23                2021                   12                11.1
24                2022                    1                11.1
25                2022                    2                11.1
26                2022                    3                11.1
27                2022                    4                 9.3
28                2022                    5                 9.3
29                2022                    6                 9.3
30                2022                    7                 8.7
31                2022                    8                 8.7
32                2022                    9                 8.7
33                2022                   10                 7.9
34                2022                   11                 7.9
35                2022                   12                 7.9
36                2023                    1                 8.8
37                2023                    2                 8.8
38                2023                    3                 8.8
39                2023                    4                 8.0
40                2023                    5                 8.0
41                2023                    6                 8.0
42                2023                    7                 8.0
43                2023                    8                 0.0
44                2023                    9                 0.0
45                2023                   10                 0.0
46                2023                   11                 0.0
47                2023                   12                 0.0

In [22]:
# Agrupa os DataFrames usando as colunas 'year' e 'month'
df = pd.merge(df_codification, df_dollar, left_on=["year", "month"], right_on=["year_dollar", "month_dollar"], how="left")

# Remove das colunas de ano e mês duplicadas após a junção
df.drop(columns=["year_dollar", "month_dollar"], inplace=True)

# Agrupa os DataFrames de empregabilidade e resultado usando a coluna 'year' e 'month'
df = df.merge(df_employability, left_on=["year", "month"], right_on=["year_employability", "month_employability"], how="left")

# Remove as colunas de ano e mês duplicadas após a junção
df.drop(columns=["year_employability", "month_employability"], inplace=True)

# Verifica o novo dataframe
df

date  weekday_name  sku  unit_price  shipment_type  \
0       2020-01-01             6  549     1099.99              0   
1       2020-01-01             6  669      413.99              0   
2       2020-01-01             6  745      949.99              0   
3       2020-01-01             6  129     2949.99              0   
4       2020-01-01             6  697      657.99              0   
...            ...           ...  ...         ...            ...   
780105  2023-07-03             1   48      504.88              0   
780106  2023-07-03             1  182     1717.95              0   
780107  2023-07-03             1  109     1659.88              0   
780108  2023-07-03             1  184      909.96              0   
780109  2023-07-03             1  446      619.98              1   

        anchor_category  product_department  product_category  origin_country  \
0                    24                   3                11               0   
1                    13                   4                 9               0   
2                    10                   4                12               0   
3                    24                   4                11               0   
4                    17                   4                 9               0   
...                 ...                 ...               ...             ...   
780105                2                   4                 9               0   
780106               23                   4                11               0   
780107                5                   4                 9               0   
780108                9                   4                12               0   
780109                1                   3                12               1   

        process_costing  ...  winning_price  revenue  items_sold  \
0                     0  ...        1488.07     0.00           0   
1                     1  ...         392.58     0.00           0   
2                     1  ...        1110.69   981.02           1   
3                     1  ...        3457.04     0.00           0   
4                     1  ...         529.00     0.00           0   
...                 ...  ...            ...      ...         ...   
780105                1  ...         339.00     0.00           0   
780106                1  ...        1717.95     0.00           0   
780107                1  ...        1093.68     0.00           0   
780108                1  ...         899.96     0.00           0   
780109                1  ...         659.98     0.00           0   

        avg_website_visits_last_week  stock_qty  month  day  year  \
0                          17.285714          0      1    1  2020   
1                          48.857143         56      1    1  2020   
2                          22.714286          0      1    1  2020   
3                           2.800000          0      1    1  2020   
4                           0.000000          0      1    1  2020   
...                              ...        ...    ...  ...   ...   
780105                     41.000000         61      7    3  2023   
780106                    154.142857          1      7    3  2023   
780107                    138.000000         16      7    3  2023   
780108                     32.714286         20      7    3  2023   
780109                     19.000000        136      7    3  2023   

        value_dollar  rate_employability  
0             4.0949                12.4  
1             4.0949                12.4  
2             4.0949                12.4  
3             4.0949                12.4  
4             4.0949                12.4  
...              ...                 ...  
780105        4.8222                 8.0  
780106        4.8222                 8.0  
780107        4.8222                 8.0  
780108        4.8222                 8.0  
780109        4.8222                 8.0  

[780110 rows x 22 columns]

In [23]:
# Verifica se há dados nulos no dataframe
df.isnull().sum()

date                            0
weekday_name                    0
sku                             0
unit_price                      0
shipment_type                   0
anchor_category                 0
product_department              0
product_category                0
origin_country                  0
process_costing                 0
sku_color                       0
price_status                    0
winning_price                   0
revenue                         0
items_sold                      0
avg_website_visits_last_week    0
stock_qty                       0
month                           0
day                             0
year                            0
value_dollar                    0
rate_employability              0
dtype: int64

## Random Forest

### Teste do modelo com a coluna `winning_price` preenchida

No notebook _Modelagem_ o grupo escolheu o algoritmo de Random Forest como o modelo candidato para o desenvolvimento do projeto. O treinamento de tal modelo foi realizado no notebook citado sem a coluna de `winning_price` preenchida, haja vista que não tínhamos conseguido fazer o preenchimento dela com a base enviada pelo parceiro. Sendo assim, após preenchermos a coluna `winning_price`, resolvemos treinar novamente o Random Forest para observar suas métricas:


In [ ]:
# Exclui as colunas "date" e "revenue" do dataframe, pois o modelo não deve ser treinado com elas
df.drop(columns=['date', 'revenue'], inplace=True)

##### Preparação

Utilizando o pacote sklearn foi feito o treinamento do modelo, separando 20% dos dados para teste.

In [ ]:
# Define a coluna que se quer prever
X = df.drop('items_sold', axis=1)
y = df['items_sold']

# Define os dados de treino e os de teste
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

##### Treinamento

In [ ]:
# Evoca o algoritmo Random Forest de regressão
random_forest = RandomForestRegressor(n_estimators=100, random_state=42)

# Treina o modelo
random_forest.fit(X_train, y_train)

RandomForestRegressor(random_state=42)

##### Teste

In [ ]:
# Realiza as previsões no conjunto de teste
y_pred = random_forest.predict(X_test)

# Calcula as métricas de avaliação
mse = mean_squared_error(y_test, y_pred)
mae = mean_absolute_error(y_test, y_pred)
rmse = sqrt(mse)
r2 = r2_score(y_test, y_pred)

# Imprime as métricas
print("Mean Squared Error (MSE):", mse)
print("Mean Absolute Error (MAE):", mae)
print("Root Mean Squared Error (RMSE):", rmse)
print("R-squared (R²):", r2)

Mean Squared Error (MSE): 11.448398912332877
Mean Absolute Error (MAE): 1.5389725166963635
Root Mean Squared Error (RMSE): 3.383548272499282
R-squared (R²): 0.6859114017884282


### Testes de relação entre features e métricas

Aqui serão conduzidos testes com o objetivo de avaliar no modelo Random Forest, que foi o que apresentou melhores resultados até agora, o impacto da exclusão do período da Black Friday, dado seu notório pico de vendas discrepante em relação ao restante do ano, bem como da inclusão de variáveis externas, como a cotação do dólar e a taxa de empregabilidade, nas métricas de desempenho. Portanto, serão avaliadas quatro combinações desses dois fatores: 

* Exclusão dos dias de **20 à 30** de Novembro e mantendo as colunas *value_dollar* e *rate_employability*

In [ ]:
# Exclui os dados do período de 20 a 30 de novembro
df_case1 = df.copy()
exclude_column = (df_case1['month'] == 11) & (df_case1['day'] >= 20) & (df_case1['day'] <= 30)
df_case1 = df_case1[~exclude_column ]

# Define a coluna que se quer prever
X_case1 = df_case1.drop('items_sold', axis=1)
y_case1 = df_case1['items_sold']

# Define os dados de treino e os de teste
X_train_case1, X_test_case1, y_train_case1, y_test_case1 = train_test_split(X_case1, y_case1, test_size=0.2, random_state=42)

# Evoca o algoritmo Random Forest de regressão
random_forest_case1 = RandomForestRegressor(n_estimators=100, random_state=42)

# Treina o modelo
random_forest_case1.fit(X_train_case1, y_train_case1)

# Realiza as previsões no conjunto de teste
y_pred_case1 = random_forest_case1.predict(X_test_case1)

# Calcula as métricas de avaliação
mse_case1 = mean_squared_error(y_test_case1, y_pred_case1)
mae_case1 = mean_absolute_error(y_test_case1, y_pred_case1)
rmse_case1 = sqrt(mse_case1)
r2_case1 = r2_score(y_test_case1, y_pred_case1)

# Imprime as métricas
print("Caso 1 - Mean Squared Error (MSE):", mse_case1)
print("Caso 1 - Mean Absolute Error (MAE):", mae_case1)
print("Caso 1 - Root Mean Squared Error (RMSE):", rmse_case1)
print("Caso 1 - R-squared (R²):", r2_case1)

KeyboardInterrupt: 

* Exclusão dos dias de **20 à 30** de Novembro e excluindo as colunas *value_dollar* e *rate_employability*

In [ ]:
# Exclui os dados do período de 20 a 30 de novembro e as colunas de dólar e empregabilidade
df_case2 = df.copy()
condicao_exclusao = (df_case2['month'] == 11) & (df_case2['day'] >= 20) & (df_case2['day'] <= 30)
df_case2 = df_case2[~condicao_exclusao]
df_case2.drop(columns=['value_dollar', 'rate_employability'], inplace=True)

# Define a coluna que se quer prever
X_case2 = df_case2.drop('items_sold', axis=1)
y_case2 = df_case2['items_sold']

# Define os dados de treino e os de teste
X_train_case2, X_test_case2, y_train_case2, y_test_case2 = train_test_split(X_case2, y_case2, test_size=0.2, random_state=42)

# Evoca o algoritmo Random Forest de regressão
random_forest_case2 = RandomForestRegressor(n_estimators=100, random_state=42)

# Treina o modelo
random_forest_case2.fit(X_train_case2, y_train_case2)

# Realiza as previsões no conjunto de teste
y_pred_case2 = random_forest_case2.predict(X_test_case2)

# Calcula as métricas de avaliação
mse_case2 = mean_squared_error(y_test_case2, y_pred_case2)
mae_case2 = mean_absolute_error(y_test_case2, y_pred_case2)
rmse_case2 = sqrt(mse_case2)
r2_case2 = r2_score(y_test_case2, y_pred_case2)

# Imprime as métricas
print("Caso 2 - Mean Squared Error (MSE):", mse_case2)
print("Caso 2 - Mean Absolute Error (MAE):", mae_case2)
print("Caso 2 - Root Mean Squared Error (RMSE):", rmse_case2)
print("Caso 2 - R-squared (R²):", r2_case2)

Caso 2 - Mean Squared Error (MSE): 9.012915251557827
Caso 2 - Mean Absolute Error (MAE): 1.4847787424084737
Caso 2 - Root Mean Squared Error (RMSE): 3.002151770240443
Caso 2 - R-squared (R²): 0.7209042579619473


* Exclusão dos dias de **10 à 30** de Novembro e das colunas *value_dollar* e *rate_employability*

In [ ]:
# Exclui os dados do período de 10 a 30 de novembro e das coluns de dólar e empregabilidade
df_case3 = df.copy()
exclude_column = (df_case3['month'] == 11) & (df_case3['day'] >= 10) & (df_case3['day'] <= 30)
df_case3 = df_case3[~exclude_column]
df_case3.drop(columns=['value_dollar', 'rate_employability'], inplace=True)

# Define a coluna que se quer prever
X_case3 = df_case3.drop('items_sold', axis=1)
y_case3 = df_case3['items_sold']

# Define os dados de treino e os de teste
X_train_case3, X_test_case3, y_train_case3, y_test_case3 = train_test_split(X_case3, y_case3, test_size=0.2, random_state=42)

# Evoca o algoritmo Random Forest de regressão
random_forest_case3 = RandomForestRegressor(n_estimators=100, random_state=42)

# Treina o modelo
random_forest_case3.fit(X_train_case3, y_train_case3)

# Realiza as previsões no conjunto de teste
y_pred_case3 = random_forest_case3.predict(X_test_case3)

# Calcula as métricas de avaliação
mse_case3 = mean_squared_error(y_test_case3, y_pred_case3)
mae_case3 = mean_absolute_error(y_test_case3, y_pred_case3)
rmse_case3 = sqrt(mse_case3)
r2_case3 = r2_score(y_test_case3, y_pred_case3)

# Imprime as métricas
print("Caso 3 - Mean Squared Error (MSE):", mse_case3)
print("Caso 3 - Mean Absolute Error (MAE):", mae_case3)
print("Caso 3 - Root Mean Squared Error (RMSE):", rmse_case3)
print("Caso 3 - R-squared (R²):", r2_case3)

Caso 3 - Mean Squared Error (MSE): 8.460861905566178
Caso 3 - Mean Absolute Error (MAE): 1.4746283923009884
Caso 3 - Root Mean Squared Error (RMSE): 2.9087560752951043
Caso 3 - R-squared (R²): 0.72401428481432


* Exclusão dos dias de **10 à 30** de Novembro e mantendo as colunas *value_dollar* e *rate_employability*

In [ ]:
# Exclui os dados do período de 10 a 30 de novembro
df_case4 = df.copy()
exclude_column = (df_case4['month'] == 11) & (df_case4['day'] >= 10) & (df_case4['day'] <= 30)
df_case4 = df_case4[~exclude_column]

# Definição da coluna que se quer prever
X_case4 = df_case4.drop('items_sold', axis=1)
y_case4 = df_case4['items_sold']

# Definição dos dados de treino e os de teste
X_train_case4, X_test_case4, y_train_case4, y_test_case4 = train_test_split(X_case4, y_case4, test_size=0.2, random_state=42)

# Evocação do algoritmo Random Forest de regressão
random_forest_case4 = RandomForestRegressor(n_estimators=100, random_state=42)

# Treinamento do modelo
random_forest_case4.fit(X_train_case4, y_train_case4)

# Realiza as previsões no conjunto de teste
y_pred_case4 = random_forest_case4.predict(X_test_case4)

# Calcula as métricas de avaliação
mse_case4 = mean_squared_error(y_test_case4, y_pred_case4)
mae_case4 = mean_absolute_error(y_test_case4, y_pred_case4)
rmse_case4 = sqrt(mse_case4)
r2_case4 = r2_score(y_test_case4, y_pred_case4)

# Imprime as métricas
print("Caso 4 - Mean Squared Error (MSE):", mse_case4)
print("Caso 4 - Mean Absolute Error (MAE):", mae_case4)
print("Caso 4 - Root Mean Squared Error (RMSE):", rmse_case4)
print("Caso 4 - R-squared (R²):", r2_case4)

Ao analisarmos os resultados das quatro combinações, é evidente que o Caso 4 se destacou ao apresentar as melhores métricas para o contexto em questão. Nesse cenário, o período que foi excluido inclui a Black Friday e abrange o dia 10 de Novembro ao dia 30 de Novembro. Essa escolha se justifica pelo fato de que o número de vendas começa a se alterar no início do mês, mesmo que a Black Friday em si ocorra por volta do dia 25. Essa influência precoce na precisão do modelo é significativa.

Além disso, a inclusão das variáveis externas no dataframe, que abrangem a variação do dólar e a taxa de empregabilidade, também contribuiu para um leve aumento no valor de R². Essa métrica é de suma importância e está no centro das nossas considerações.

Sendo assim, iremos considerar esse caso para aplicarmos os hiperparâmetros.

In [ ]:
# Baixa o dataframe com os dados do melhor modelo na pasta do notebook atual
df_case4.to_csv('mobly_data_official.csv', index=False)